### ISSUE TITLE
Allow temporary / ephemeral overrides of settings

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#14084]
### Problem

Games may want to change settings to provide a better user experience. For example, many games don't want time to progress, and hence set `time_speed` to `0`. (The alternative "hack" is to set the time of day for each player in a globalstep, but this produces unnecessary traffic and may be janky depending on lag and the value of `time_speed`, though usually it is good enough.)

Currently, touching settings at all is **a massive footgun**:

* Many modders forget to reset settings at all, overwriting global user settings, "gaslighting" the user.
* Even if modders restore settings, settings will remain overwritten in case of any crash (including Lua crashes).
* The last point can be alleviated using even more hacks (for example, you could try restoring settings in `minetest.error_handler` or `__gc` metamethods of objects which are collected at shutdown). I don't think I've seen anyone but me ever do this, and even then this is still fragile in case of even more abrupt shutdowns (such as power outages).
* You can then add "remembering" settings and "restoring" them when the game is rejoined, but well, that requires the user to rejoin the game to have their settings restored.

### Solutions

Allow to "temporarily" set settings *such that the modifications are not persisted*. I'm not sure what the best API for this would be, but I'm thinking something like `minetest.settings:set_ephemeral(key, value)`. Ephemeral settings will not be persisted; instead, the last non-ephemeral value assigned to `key` would be persisted.

### Alternatives

* Provide alternative APIs to modify these settings. It seems unlikely to me that anyone put in the work for this (going over all the settings and conceiving alternative Lua APIs), and even then it seems likely that they would neglect some settings.
* Expand documentation, warning modders *not to set settings unless they also plan to reset them*, and also instructing modders to *please inform their users about how they are messing with their settings*.
* Have the engine "notify" the user when a mod changes their global settings.

### Additional context

Quite a few packages setting `time_speed`, most often not making any effort to reset it, can be found [on ContentDB](https://content.minetest.net/zipgrep/f71fad64-3d68-466b-bcad-b530f495a8bb/). Most of these will break the user's global settings. (Note also: Many of these just get `time_speed`, which is fine.)

Modifying settings is pretty powerful. For example, in a singleplayer game, it lets you temporarily (!) change the client font by setting the font paths to paths of fonts in your game/mod folders, then restoring them afterwards. But of course, if the game crashes, the player will be stuck with the font. These are only two settings out of many, many settings which can be modified at load time and still take an effect.

**Comment by rubenwardy** [comment]
Alternative: #6966

**Comment by appgurueu** [comment]
> Alternative: #6966

Right, I should have mentioned that to. I briefly considered that, but it's less powerful than "ephemeral" settings: World-specific settings can trivially be implemented by a mod in terms of "ephemeral" settings, but not the other way around.

**Comment by Desour** [comment]
This is one of the bugs that annoy me the most, because I can't try out new mods and subgames without the fear that they mess around with my settings.

---

A settings set api that modifies global settings is inappropriate, especially if the server is setting client settings (which happens to work in singleplayer, see e.g. your font example).
It should rather modify a separate settings layer, may it be your suggested ephemeral settings, or world settings. The only difference I see is that one is persistent and the other not. The persistent world setting would have the advantage that the `set` chatcommand would still work.

---

>World-specific settings can trivially be implemented by a mod in terms of "ephemeral" settings, but not the other way around.

(You could make a mod that resets all settings on startup or on shutdown.)

**Comment by Warr1024** [comment]
Compared to #6966, the implications of this approach are interesting.

In theory, it _should_ be possible to emulate #6966 by just having a mod/game save/load its settings in a world-specific place (mod_storage) and apply them as ephemeral overrides on startup automatically.  But many mods only ever read their settings on startup and require a restart to receive updated config, so unless this is done in builtin or there's some way to guarantee loading order (dependencies only would work insofar as we can trust the ecosystem to standardize around a single implementation or not-forever-expanding set, which is nearly unprecedented) that probably would be full of holes in practice.

Assuming a way to emulate #6966 is found, this seems like it would be a superset of functionality.  This seems particularly well suited to some applications, like https://gitlab.com/sztest/szutilpack/-/tree/master/szutil_clocksync.
